tech_reports/LLMs/llms.md

```diff
@@ -51,13 +130,292 @@ Authors:
+    per_core_M = math.ceil(m / (tile_size * grid_size[1]))
+    per_core_N = math.ceil(n / (tile_size * grid_size[0]))
+
+    out_subblock_h = 1
+    out_subblock_w = get_out_subblock_w(per_core_N, out_subblock_h)
+
+    if in0_block_w is None:
+        in0_block_w = min(4, max(1, k // (tile_size * grid_size[0])))
+
+    return ttnn.MatmulMultiCoreReuseMultiCastProgramConfig(
+        compute_with_storage_grid_size=grid_size,
+        in0_block_w=in0_block_w,
+        out_subblock_h=out_subblock_h,
+        out_subblock_w=out_subblock_w,
+        per_core_M=per_core_M,
+        per_core_N=per_core_N,
+        transpose_mcast=False,
+        fused_activation=fused_activation,
+        fuse_batch=fuse_batch,
+    )
+
+
+_, _, m, k = ff_in.shape
+n = hidden_dim // num_devices
+pc1 = matmul_config(
+    m=m, k=k, n=n, grid_size=(8, 8)
+)
+
+k, n = n, k  # Since FF1 is up projection and FF2 is down projection
+pc1 = matmul_config(
+    m=m, k=k, n=n, grid_size=(8, 8)
+)
+```
+
+
+#### 3. FF1/FF3 matmul
+The first set of operations in the MLP are:
+```py
+w1_out = FF1(x)
+w3_out = FF3(x)
+```
+Based on the program configs we computed beforehand, we perform the FF1/FF3 matmuls, making sure that the ouputs are L1 sharded in in decode mode, and interleaved in DRAM if in prefill mode. For the `compute_kernel_config`, we use `ttnn.MathFidelity.HiFi2` to retain accuracy while still being performance. Using `ttnn.MathFidelity.HiFi4` instead, would mean that this matmul would become compute bound.
+
+```py
+compute_kernel_config_hifi2 = ttnn.WormholeComputeKernelConfig(
+    math_fidelity=ttnn.MathFidelity.HiFi2,
+    math_approx_mode=False,
+    fp32_dest_acc_en=False,
+    packer_l1_acc=True,
+)
+
+w1_out = ttnn.linear(
+    ff_in,
+    w1,
+    compute_kernel_config=args.compute_kernel_config_hifi2,
+    core_grid=ttnn.CoreGrid(y=8, x=8) if not pc_1 else None,
+    dtype=ttnn.bfloat16,
+    program_config=pc_1,
+    memory_config=ttnn.L1_WIDTH_SHARDED_MEMORY_CONFIG if mode == "decode" else ttnn.DRAM_MEMORY_CONFIG,
+)
+
+w3_out = ttnn.linear(
+    ff_in,
+    w3,
+    compute_kernel_config=args.compute_kernel_config_hifi2,
+    core_grid=ttnn.CoreGrid(y=8, x=8) if not pc_1 else None,
+    dtype=ttnn.bfloat16,
+    program_config=pc_1,
+    memory_config=ttnn.L1_WIDTH_SHARDED_MEMORY_CONFIG if mode == "decode" else ttnn.DRAM_MEMORY_CONFIG,
+)
+```
+
+#### 3.1 FF1/FF3 Matmul with 2D Weight Fracturing
+
+In the case of TG systems, where have have access to a 2D device mesh, we can leverage 2D weight fracturing. For a weight tensor with shape `[1, 1, K, N]`, using 2D weight fracturing on a `(8, 4)` device mesh, the resulting shape on each device would be: `[1, 1, K / 4, N / 8]`. In other words, the inner dimension (K) of the matmul is spread out across 4 devices, and to complete the entire matmul operation, a reduction step across the partials is necessary. We do this using an all-reduce operation along the 4 devices in `cluster_axis=1` of the device mesh.
```

Comment:
Johanna is talking about 2D sharding in "multi-device" section. In this section, let's talk about what happens in each device. Let's add this sharding, allreduce, allgather discussion to Johanna's section.